core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationEndpointFilter.java

```diff
@@ -15,33 +15,252 @@
+import org.springframework.util.StringUtils;
 import org.springframework.web.filter.OncePerRequestFilter;
-
-import javax.servlet.FilterChain;
-import javax.servlet.ServletException;
-import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
-import java.io.IOException;
+import org.springframework.web.util.UriComponentsBuilder;
 
 /**
  * @author Joe Grandja
+ * @author Paurav Munshi
+ * @since 0.0.1
  */
 public class OAuth2AuthorizationEndpointFilter extends OncePerRequestFilter {
+
 	private Converter<HttpServletRequest, OAuth2AuthorizationRequest> authorizationRequestConverter;
 	private RegisteredClientRepository registeredClientRepository;
 	private OAuth2AuthorizationService authorizationService;
 	private StringKeyGenerator codeGenerator;
+	private RedirectStrategy authorizationRedirectStrategy;
+	private RequestMatcher authorizationEndpiontMatcher;
+
+	private static final String DEFAULT_ENDPOINT = "/oauth2/authorize";
+
+	private static final OAuth2Error CLIENT_ID_ABSENT_ERROR = new OAuth2Error(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2AuthorizationServerMessages.REQUEST_MISSING_CLIENT_ID, null);
+	private static final OAuth2Error REDIRECT_URI_REQUIRED = new OAuth2Error(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2AuthorizationServerMessages.REDIRECT_URI_MANDATORY_FOR_CLIENT, null);
+	private static final OAuth2Error INVALID_REDIRECT_URI_REQUESTED = new OAuth2Error(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2AuthorizationServerMessages.REQUESTED_REDIRECT_URI_INVALID, null);
+	private static final OAuth2Error CLIENT_ID_NOT_FOUND_ERROR = new OAuth2Error(OAuth2ErrorCodes.ACCESS_DENIED, OAuth2AuthorizationServerMessages.CLIENT_ID_NOT_FOUND, null);
+	private static final OAuth2Error USER_NOT_AUTHENTICATED_ERROR = new OAuth2Error(OAuth2ErrorCodes.ACCESS_DENIED, OAuth2AuthorizationServerMessages.USER_NOT_AUTHENTICATED, null);
+	private static final OAuth2Error AUTHZ_CODE_NOT_SUPPORTED_ERROR = new OAuth2Error(OAuth2ErrorCodes.ACCESS_DENIED, OAuth2AuthorizationServerMessages.CLIENT_ID_UNAUTHORIZED_FOR_CODE, null);
+	private static final OAuth2Error RESPONSE_TYPE_NOT_FOUND_ERROR = new OAuth2Error(OAuth2ErrorCodes.UNSUPPORTED_RESPONSE_TYPE, OAuth2AuthorizationServerMessages.RESPONSE_TYPE_MISSING_OR_INVALID, null);
+
+
+
+	public OAuth2AuthorizationEndpointFilter() {
+		authorizationEndpiontMatcher = new AntPathRequestMatcher(DEFAULT_ENDPOINT);
+		authorizationRequestConverter  = new OAuth2AuthorizationRequestConverter();
+		codeGenerator  = new AuthorizationCodeKeyGenerator();
+		authorizationRedirectStrategy  = new DefaultRedirectStrategy();
+	}
 
 	@Override
 	protected void doFilterInternal(HttpServletRequest request,
 			HttpServletResponse response, FilterChain filterChain)
 			throws ServletException, IOException {
 
+		RegisteredClient client = null;
+		OAuth2AuthorizationRequest authorizationRequest = null;
+		OAuth2Authorization authorization = null;
+
```

Comment:
I would have expected to see the use of `authorizationEndpointMatcher` as the very first call here. However, it's not being used at all. Please review [4.1.1.  Authorization Request](https://tools.ietf.org/html/rfc6749#section-4.1.1). At a minimum, there should be a `RequestMatcher` that matches on a **valid Authorization Request**. A valid authorization request should at least match on the configured endpoint path (using `AntPathRequestMatcher`) and the required parameters `response_type=code`, `client_id=*`. This is a common pattern used throughout Spring Security, where `Filter` use a `RequestMatcher` to filter out request or process them.

Reply:
I have used the RequestMatcher in the shouldNotFilter method. That method is called by OncePerRequestFilter.doFilter method effectively calling it before doFilterInternal. I thought it is much more natural to filter this way for checking endpoint. And I also think that validity of the query parameters is not supposed to be done in RequestMatcher because it is only responsible to check if filter should be invoked for that request or not. Once invoked, I do not think RequestMatcher should validate the request and therefore I have those validations in separate methods. Let me know if it is not to be done in that way.

Reply:
> I have used the RequestMatcher in the shouldNotFilter method

Sorry  missed that.

> I also think that validity of the query parameters is not supposed to be done in RequestMatcher

A valid Authorization Request requires `response_type=code` and `client_id != empty` so both of these need to be applied in `RequestMatcher`

Reply:
> A valid Authorization Request requires response_type=code and client_id != empty so both of these need to be applied in RequestMatcher

So here is where my confusion lies. `RequestMatcher `by its nature, it seems, is supposed to decide whether request is supported by that filter or not. So if the `RequestMatcher `does not find a match it will silently allow the request to proceed to `FilterChain`. But if we look at first line of [4.1.2.1 Error Response](https://tools.ietf.org/html/rfc6749#section-4.1.2.1), it says :- 

> If the request fails due to a missing, invalid, or mismatching
>    redirection URI, or if the client identifier is missing or invalid,
>    the authorization server SHOULD inform the resource owner of the
>    error

So if we are putting` response_type=code and client_id != empty` in `RequestMatcher` it will not send the error response back to resource owner's user-agent. So how are supposed to handle this scenario if we want to perform this validation in `RequestMatcher`. I think that in future if Implicit grant is going to be implemented using a different filter then we can add `response_type=code` in `RequestMatcher` but I would still not think of validating client id in `ReqeustMatcher`. 

Can you please provide some more light here ?

Reply:
At a minimum I have added the check for response_type in RequestMatcher. So if the response_type is blank or if its not code then this filter will not be invoked and filter chain will proceeded with. Tests have been added to check the same. 

Can you please review the PR ?